docs/spec-files/encryption.md

```diff
@@ -107,6 +107,9 @@ When Bob receives the envelope, the unpacking process on his end MUST resolve th
 
 Once resolved, the unpacker will then execute ECDH-1PU key derivation using this key and Bob's own recipient key found in the envelope's `recipients[0]` to unwrap the content encryption key.
 
+#### Note about Messages without DIDs
+In requests where the DIDs have not yet been exchanged (eg: OOB invitations or the actual DID exchange request), agents MAY send these requests in plaintext. If an agent accepts only encrypted messages, only then an encrypted envelope MAY be used. Since the DIDs are still not exchanged in this specific case, the `kid`/`skid` values SHOULD be the `did:key` representation of the key. Once DIDs are exchanged, only encrypted messages MUST be used and `kid`/`skid` values MUST adhere to the DID's `KeyAgreement` section as mentioned earlier.
+
```

Comment:
@Baha-sk Could you provide an example flow?

Reply:
@awoie, sorry for the delay, I had some issues with my bdd tests and backwards compatibility between DIDComm V1 and V2, but here is an example scenario built using afgo (extracted from our bdd tests run):

a sample didexchange request is the following:
```{"@type":"https://didcomm.org/didexchange/1.0/request","@id":"f155ff9a-994a-4718-8ce3-b88abceac9b7","label":"Bob","~thread":{"pthid":"1df76821-1609-4dc8-a1f5-8bdb13ca3c5e"},"did":"did:peer:1zQmeHCBHs1ZD7PHPLY6366RR3ryGxS6Dz7PxFid2TmmYA7S","did_doc~attach":{"mime-type":"application/json","lastmod_time":"0001-01-01T00:00:00Z","data":{"base64":"eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvbnMvZGlkL3YxIl0sImlkIjoiZGlkOnBlZXI6MXpRbWVIQ0JIczFaRDdQSFBMWTYzNjZSUjNyeUd4UzZEejdQeEZpZDJUbW1ZQTdTIiwidmVyaWZpY2F0aW9uTWV0aG9kIjpbeyJjb250cm9sbGVyIjoiIiwiaWQiOiIja2V5LTEiLCJwdWJsaWNLZXlCYXNlNTgiOiIyM1c3dXVua0ViY2VjZkJzRldhbjI1WkV3dHVCVVVhVURzdkNUcU1mSkxBWSIsInR5cGUiOiJFZDI1NTE5VmVyaWZpY2F0aW9uS2V5MjAxOCJ9LHsiY29udHJvbGxlciI6IiIsImlkIjoiI2tleS0yIiwicHVibGljS2V5SndrIjp7Imt0eSI6IkVDIiwia2lkIjoiVVRQeVlVSHhaclo3SHExV1M2eGRudlI3OGRWRnVJbDhEMkpCbmFNcXp4RSIsImNydiI6IlAtMzg0IiwieCI6IlFUVk1VMnR3MlRKZzZ3dmFhZlVEc19YeEFQUTRIeVFNS0xBdDNwY25lbXRjdFRqOF9NNFhVRi1OdnhoWmwyYlUiLCJ5IjoiZXAxaWlrSjQxXzcxMUs0VlJoLWVlRkpJTXVIQ1hIVDU1NmJ2WXY2a3FTeU1lV19CYVExTnFiNUZCRGxBODFSdyJ9LCJ0eXBlIjoiSnNvbldlYktleTIwMjAifV0sInNlcnZpY2UiOlt7ImlkIjoiZDE2MzBiMjQtNGFmZi00MjdmLThmYzQtZTBiZTgzMzJjOTgyIiwicHJpb3JpdHkiOjAsInJlY2lwaWVudEtleXMiOlsiZGlkOnBlZXI6MXpRbWVIQ0JIczFaRDdQSFBMWTYzNjZSUjNyeUd4UzZEejdQeEZpZDJUbW1ZQTdTI2tleS0yIl0sInJvdXRpbmdLZXlzIjpbXSwic2VydmljZUVuZHBvaW50IjoiaHR0cDovL2xvY2FsaG9zdDo1NzE0MCIsInR5cGUiOiJESURDb21tTWVzc2FnaW5nIn1dLCJhdXRoZW50aWNhdGlvbiI6WyIja2V5LTEiXSwiYXNzZXJ0aW9uTWV0aG9kIjpbIiNrZXktMSJdLCJrZXlBZ3JlZW1lbnQiOlsiI2tleS0yIl0sImNyZWF0ZWQiOiIyMDIxLTA5LTIwVDE4OjAzOjE1LjUzNjMwNS0wNDowMCIsInVwZGF0ZWQiOiIyMDIxLTA5LTIwVDE4OjAzOjE1LjUzNjMwNS0wNDowMCJ9"}}}```

data.base64 value above decoded: 
```
{
	"@context": ["https://www.w3.org/ns/did/v1"],
	"id": "did:peer:1zQmeHCBHs1ZD7PHPLY6366RR3ryGxS6Dz7PxFid2TmmYA7S",
	"verificationMethod": [{
		"controller": "",
		"id": "#key-1",
		"publicKeyBase58": "23W7uunkEbcecfBsFWan25ZEwtuBUUaUDsvCTqMfJLAY",
		"type": "Ed25519VerificationKey2018"
	}, {
		"controller": "",
		"id": "#key-2",
		"publicKeyJwk": {
			"kty": "EC",
			"kid": "UTPyYUHxZrZ7Hq1WS6xdnvR78dVFuIl8D2JBnaMqzxE",
			"crv": "P-384",
			"x": "QTVMU2tw2TJg6wvaafUDs_XxAPQ4HyQMKLAt3pcnemtctTj8_M4XUF-NvxhZl2bU",
			"y": "ep1iikJ41_711K4VRh-eeFJIMuHCXHT556bvYv6kqSyMeW_BaQ1Nqb5FBDlA81Rw"
		},
		"type": "JsonWebKey2020"
	}],
	"service": [{
		"id": "d1630b24-4aff-427f-8fc4-e0be8332c982",
		"priority": 0,
		"recipientKeys": ["did:peer:1zQmeHCBHs1ZD7PHPLY6366RR3ryGxS6Dz7PxFid2TmmYA7S#key-2"],
		"routingKeys": [],
		"serviceEndpoint": "http://localhost:57140",
		"type": "DIDCommMessaging"
	}],
	"authentication": ["#key-1"],
	"assertionMethod": ["#key-1"],
	"keyAgreement": ["#key-2"],
	"created": "2021-09-20T18:03:15.536305-04:00",
	"updated": "2021-09-20T18:03:15.536305-04:00"
}
```

this whole request (not just data.base64 value) is the payload being packaged (encrypted) prior to being sent by Alice to Bob. As you can see, the above payload contains Alice's DID document that Bob is about to receive encrypted in an envelope.

the resulting envelope would be the following serialized JWE message (since Bob has 1 key only in this example):
```
eyJhbGciOiJFQ0RILTFQVStBMjU2S1ciLCJhcHUiOiJaR2xrT210bGVUcDZPREpNYTI1MWNuSnlXSGRaYldGQ1ZXVnhVRUo2WW0xdlpHUkhkemQwVTFOUWEwSklOMjFyYVZscFpUVm9lbFUyZDNSa1p6ZzBTRGgwWmxVNU16VlpNVFYwT0VkTVlnIiwiYXB2IjoiekZhX0NjSDI5UVRmekdhOGdjQjFVZ1EtUk05TlRwT09lRHlmWUMzRW9qMCIsImN0eSI6ImFwcGxpY2F0aW9uL2RpZGNvbW0tcGxhaW4ranNvbiIsImVuYyI6IkEyNTZDQkMtSFM1MTIiLCJlcGsiOnsia3R5IjoiRUMiLCJjcnYiOiJQLTM4NCIsIngiOiJNZ3dLejU3WjNqSXBlZWZvOW5jdXFiZ0sxSFRGLTVWV0Qta2JVdGxHQjliRTBtcWUwZmx4ek1TMUpKYm9iRlFoIiwieSI6IlNINVloT0hoTl9NNE9wNVZpbFl0VG9IV3hySmtULWpIekVGbWdxRnIxREh0b2xRTTZOeXMyZzBFTVplV053MzAifSwia2lkIjoiZGlkOmtleTp6ODJMa25hbmVQRTNUak5iNE1DTlNGb2FOY0djdkt4VXBoVVhjbUVabUtibjc1YmdXZTR6SzdnUlhCZDh1WjZRajNMa2cyayIsInNraWQiOiJkaWQ6a2V5Ono4MkxrbnVycnJYd1ltYUJVZXFQQnpibW9kZEd3N3RTU1BrQkg3bWtpWWllNWh6VTZ3dGRnODRIOHRmVTkzNVkxNXQ4R0xiIiwidHlwIjoiYXBwbGljYXRpb24vZGlkY29tbS1lbmNyeXB0ZWQranNvbiJ9.QvbBhHNwVGBs6YeDe3v8Tu2rvGbutgsYM-sf1HYvpyExPlOW375-rJAK5CCzSsk_fzkXmvznyBUDCLvjpl1915X8PFyjFOgJ.5LLOC5hKQd_eQg2mcmX4OA._RexheGdiZdB7tOrsANNCOwqB5HfzxglSrCNLjwGKXQ9k5uBHdoC70pw3eZnS8dXMgf8xD0DqsKA7KyS-P05hTB2wArtibvf5JNiEBL_JAbUZJD4mnsm9_rPa24D6Fl-giP7J6nKPVVmn-VQs0z34F_t8SkuMVTRh9sf-8skE7ahXLpIZjqOHoLwmRBflnKtJxcvLf7SFxqauaWUfuiuOLLFkauF3btZ0MHvbYPgg2SUfLwr8X8HhzjPHR1Kip24a0hYf8821r71AlAgHc6E0pOjl4jDPjqlGSLucqBa7px84qlbVLialNKv0ve2Ed_WgpumFKCllL78ZsAG3O8TSZrlltrBjM-PrDTspgZgkryQxxDydVRKWppOjZ3WYX-c-OCn63iRqafU5_nsAVDDINdnKYV-A45bX0V2mPoVBjw-tNrUefxIbA5-mPGxHa9-k0xVeycgqBm10yRu_K1XxdQoJqnqaHpnKE8hE4EXYlR6wxK3IY0OAkB79q3eI6m4qHJYYqPpAkWh6JpN0lB7iStTrf1HZ-V_tOg5nXI797T6el80paW0r1l8jvZbvkkNoSbTtVyoSZYCwIDTnynGQJhCMc1Q9xQeEjs6T34su6Il2elLVdHztuaX6QauYpMx5nPMHYnFdu00dnVGjKhI1XXqg92JWb5WhELBScZ_XTAv-pSuuHPnYUXiCzNTF4bnejL12qJLCmxQQXd4tbi0wiKg_5rvuO2-87TlXC6-Z5mZULC7FzykezgAIaJjKJgl1X0SHUz4AEBKYzMXAxhOfZxSl1ZWq2FA2_aVxrKDSyb7gVVw8VUFM1mTeSRCHOyPPHsC9hDVkypjTBDvrkb_mwBmGovSq75nlhpA_TskWFCrHRoxLs2AFbUB4O9ci_1z3P3Z6sV6jbEeXN-NVEdfa7tJE8K8tMfbT2N8430f01m7u9s00KsnMT5IZGLhgfVioB3siT3_TP4fPa6K4IAbqKkbszVvNGOflFBUNTczSeLqYm8uSBQhSBtmps-Cup7fr4-qkJKSo2UXV_9ccYwb91OIkvT9mzs5ozdZP4iTDNAMhN_viHM0D7XaH2SCcl5vT2x7XD_MP3G2VL_sUlTcJpi6Dhn1Qfb4Or7o1QelDIxdvQqDeOTDnz5sTse645pgr8dMDhER1bM4ToqJDmYmcYICsRAe2UDSFb63_eDdXGw3Mq8eEExckzEMjPWtDenbD6vNCYFgI6NCeeb_g3TksVeFdq0aXr0FsNfgYwkpCD3eM2ZOOAt_30Yh13RgwWB7mkR5v59hBuqRC1oh2G-g6Upv_m2LTknWMfmnWzzm4s3Bsx-DThkzBaPLOJJ3fvQuCNoAVz-SHRW2M-9XX2aX9NvTwRSFaFeX0Sy3X85Pk9-ObdNafa9gRlDFN3HO6BybFkYtccAm6KyxbqFEBxkgXnEinL6tQPAZgcyGdNNQb_8kJ-G81X8FzAkOpzVADkLOd9fveMo9K-XJmmakiq2Men1b0ZdQkukN2GKgJX6Hly_uLHhrN5kXFkd_zIYkbgZk_SIcBTdbipJfkOIy3VWtZLYfQKtM4pqM9MgaKOqVJpm8n2J-304AE3EhtjmjkR_qbNK0QC3GbY_SZ1E7xZ5HTQuSaCpQazGZ24g0pDhnJy4loGx2z7JZk9zCgbqdoyQuoo6LgA82Lp4Xhv2l57RuJLqRxdS5zFPHU-kMPWjXOL2uglpVt8BTUg2-jFBa7aI6IZ8bYyng6-SmkE1_Nof8f0UxJiep_bLjOE-U5JWwzmS59muTwvB6_zkBt-z_7JSIKsggagr2RdTmJ4Le8__TLQu3srfDBk5jbEhAgaBxDS9gALJPYCoLoOHd-rVhxLQKyumFDa4lTJSi5LkfuGL6tx5b-I_UQOIEasg8BC7CdhqEz3rOYntigtM41Do-kIp5Y7fDFVyFhWKn0uZwdQQ7Nt8USQtNDWB9acTZx_DVb2v3_hJM35ppRHfa9mym7-bQmK8_ixAEGpo5sQrb8VruZ-sLAyXyb8WbbLLrn1HDF87_MGEdCwpV11hU8H_n6uEWWrGmVpSCHPGSXt33R870QXd9KDA9FlLs9RCnAbWa9WHslRcVbkVJIC2feGruGFLF9VU32qYkjgjYC04u-R8DUHkd1cY3FsuPHIOWt_DZE_u7nrAbNqFxLWAVmWzOOfPdhcbdDboOMILZjKH0amxPfg.U85-pFDJeGRYWanV3NjhoDXBPli-zf7KzWb4diaGqrk
```

The first part (content before the first `.`) is the protected headers, base64 decoding this part is shown below:
```
{
	"alg": "ECDH-1PU+A256KW",
	"apu": "ZGlkOmtleTp6ODJMa251cnJyWHdZbWFCVWVxUEJ6Ym1vZGRHdzd0U1NQa0JIN21raVlpZTVoelU2d3RkZzg0SDh0ZlU5MzVZMTV0OEdMYg",
	"apv": "zFa_CcH29QTfzGa8gcB1UgQ-RM9NTpOOeDyfYC3Eoj0",
	"cty": "application/didcomm-plain+json",
	"enc": "A256CBC-HS512",
	"epk": {
		"kty": "EC",
		"crv": "P-384",
		"x": "MgwKz57Z3jIpeefo9ncuqbgK1HTF-5VWD-kbUtlGB9bE0mqe0flxzMS1JJbobFQh",
		"y": "SH5YhOHhN_M4Op5VilYtToHWxrJkT-jHzEFmgqFr1DHtolQM6Nys2g0EMZeWNw30"
	},
	"kid": "did:key:z82LknanePE3TjNb4MCNSFoaNcGcvKxUphUXcmEZmKbn75bgWe4zK7gRXBd8uZ6Qj3Lkg2k",
	"skid": "did:key:z82LknurrrXwYmaBUeqPBzbmoddGw7tSSPkBH7mkiYie5hzU6wtdg84H8tfU935Y15t8GLb",
	"typ": "application/didcomm-encrypted+json"
}
```

Since Bob does not have Alice's DID yet, `kid`/`skid` values are `did:key` values for now. Once Bob unpackages this envelope and reads the payload, he can store the wrapped DIDdoc in his connection store for further communications. 

This is what I mean by the text in the pr where `did:key` values can be used when DIDs are not yet exchanged between agents.

Once the DIDs are exchanges,`kid`/`skid` will then have the value `did:peer: ....#key-2` corresponding to  Bob & Alice's DIDs KeyAgreement IDs. 
Example of `kid` with Alice's did:peer key Agreement ID from the DID doc above (ie Bob sending a new request) is: `did:peer:1zQmeHCBHs1ZD7PHPLY6366RR3ryGxS6Dz7PxFid2TmmYA7S#key-2`.